### DISCUSSION TITLE
Why is Quarto duplicating `.bib` files when rendering a manuscript project?

### DISCUSSION THREAD

**Discussion by arnaudgallou** [quarto-dev/quarto-cli#10510] [Q&A] [answered]
### Description

I've a simple Quarto manuscript project with the following files:

```
manuscript
├── _quarto.yml
├── index.qmd
└── references
    ├── main.bib
    └── pkrt.bib
```
    
On rendering, Quarto adds the following folders and files:

```
manuscript
├── _manuscript
│   ├── _tex
│   │   ├── index.tex
│   │   └── references
│   │       ├── main.bib
│   │       └── pkrt.bib
│   └── index.pdf
.
.
```

As you can see, there's a copy of the `.bib` files below `_tex/`. Maybe I'm missing something but what's the purpose of duplicating these files?

Also, note that I was using `keep-tex: false` in `_quarto.yml`, so I was a bit surprised that Quarto kept the `.tex` file.

**Comment by mcanouil** [comment]
Could you share a **small self-contained "working" (reproducible)** example to work with, _i.e._, a complete Quarto document or a Git repository? The goal is to make it as easy as possible for us to recreate your problem so that we can fix it: please help us help you! Thanks.

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by arnaudgallou** [reply]
Here's [an example repo](https://github.com/arnaudgallou/msproject) and the quarto check:

```
Quarto 1.5.56
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.2.0: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.41.0: OK
      Typst version 0.11.0: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.5.56
      Path: /Applications/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: (external install)
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: TinyTex
      Path: /Users/arnaudgallou/Library/TinyTeX/bin/universal-darwin
      Version: 2024

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.9.12
      Path: /usr/local/opt/python@3.9/bin/python3.9
      Jupyter: 5.7.1
      Kernels: python3

[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........OK
      Version: 4.3.2
      Path: /Library/Frameworks/R.framework/Resources
      LibPaths:
        - /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library
      knitr: 1.45
      rmarkdown: 2.27

[✓] Checking Knitr engine render......OK
```

I edited my message when you posted yours. I'm using `keep-tex: false` so I presume there's a bug somewhere anyway.

**Reply by mcanouil** [reply]
`keep-tex: false` is the default.

**Comment by mcanouil** [comment]
I think you miss-understood the purpose of the Manuscript project type and what `keep-tex` is for.

`keep-tex` is for debugging, it allows to keep the LaTeX file produced by Quarto/Pandoc to investigate it when needed, thus it is not enabled by default.

Manuscript main goal is publishing articles. `_tex` is what you need to publish your document, so Quarto is making sure all the pieces are there. `_manuscript` directory only contains outputs.

**Reply by arnaudgallou** [reply]
Yes but that doesn't answer my initial question. I was experimenting with Quarto manuscript project and used `keep-tex: false` thinking I'd avoid "polluting" the project structure with unnecessary files during the experimental phase. I'm aware that I need the `.tex` file when submitting the manuscript so this is not my issue. My question really is about the duplicated `.bib` files, because in that case I don't see the purpose of it while being a bit confusing.

**Reply by mcanouil** [reply]
How does the output pollute your project exactly?
`_manuscript` contains all the necessary things to publish. It's content is self-contained thus the bib file is also in the output directory.
This is a design choice which is more convenient that asking user to look for their bib file(s) in the whole project, then go to the output and find the LaTeX document.
I think you are confusing input and output. Quarto does not add anything outside the output directory in manuscript (besides specific temporary Quarto directory).

Try `keep-tex: true` you'll see what it does. Using your own words, this will pollute your project as it will keep a LaTeX version of your manuscript next to your manuscript and not in the output directory.

**Reply by arnaudgallou** [reply]
But why is the `.bib` file necessary when sending a `.tex` file that already uses the right csl to a journal since the `.tex` file already has all the references in it? I might have used a bad approach but I've always sent `.tex` files alone when publishing to a journal and I've never been asked to attach any `.bib` file with it.

Regarding `keep-tex`, I get the exact same thing whether I set it to `true` or `false`:

![msproject](https://github.com/user-attachments/assets/23cfbd15-20f1-419c-8dbc-134211468262)

**Reply by mcanouil** [reply]
I think here Quarto chose the more conservative approach which aopparently does not account for the framework used for the citations. This would be a feature request.
If you change the cite method to `cite-method: biblatex` you'll see that if you don't have the bib file, you have no references.

It seems `keep-tex` cannot be used inside Manuscript project but I don't know if that's intentional or a side effect of the copy of the LaTeX files to the output directory.

**Reply by arnaudgallou** [reply]
> I think here Quarto chose the more conservative approach which aopparently does not account for the framework used for the citations. This would be a feature request.

Alright. In the meantime I'll simply delete the duplicated file with:
```
post-render: rm _manuscript/_tex/references.bib
```
I'd be interested to know the exact rationale behind that feature if a maintainer reads this.

> It seems `keep-tex` cannot be used inside Manuscript project but I don't know if that's intentional or a side effect of the copy of the LaTeX files to the output directory.

That certainly the latter, which would make sense to avoid an unnecessary duplication of file. That's what confused me though because I then thought that `keep-tex` was set to true by default. I usually have invalid options being underlined but that wasn't the case for `keep-tex` so I assumed it was bugged.

**Reply by mcanouil** [reply]
> I'd be interested to know the exact rationale behind that feature if a maintainer reads this.

Clearly the rational is to allow users to use `cite-method: biblatex` (or using `natbib`) and still get a full usable output.
Now, it's not necessary to keep the bib file when the cite-method is citeproc, so as I said, this could be an enhancement.

**Reply by mcanouil** [reply]
> That certainly the latter, which would make sense to avoid an unnecessary duplication of file. > That's what confused me though because I then thought that keep-tex was set to true by default. I usually have invalid options being underlined but that wasn't the case for keep-tex so I assumed it was bugged.

As I said `keep-tex` does not output to output location, it keeps the files in place next to the Quarto document that originates them.
See <https://quarto.org/docs/output-formats/pdf-basics.html#latex-output>.

> will also create a `.tex` file alongside it that you can do subsequent processing on.

**Reply by mcanouil** [reply, marked as answer]
For reference:

- https://github.com/quarto-dev/quarto-cli/issues/10515
